### DISCUSSION TITLE
What's the goal of the TUI mode?

### DISCUSSION THREAD

**Discussion by favonia** [ocaml/dune#8994] [General]
Hi, I was curious about `dune build --watch --display tui` when I saw 3.11 is out, but then got confused about its purpose after trying it. What do I gain besides `dune build --watch`? It seems I cannot do anything more than that. I checked through [GitHub issues tagged with `tui`](https://github.com/ocaml/dune/labels/tui) and became more puzzled. It seems the leading developer of this feature is [targeting the integrated terminal inside VS Code (see #8770)](https://github.com/ocaml/dune/issues/8770), and maybe there are plans to tighten the integration with the editor. However, if we are already inside VS Code, wouldn't it be easier to use LSP instead? What's the goal of this new TUI mode within VS Code? I failed to find more information from the changelog and the codebase, and felt I must have missed something here.

---

_PS: There are also some Unicode issues: It assumes `᚛`, `―`, and `᚜` will always align, but that's not the case in my terminal font (and nothing in the Unicode standard implies they should). It also seems to assume the box drawing characters with "Ambiguous" East Asia width always have the visual width of one cell, which is unfortunately not true, either, and that's why they are "Ambiguous". However, before trying to fix the character choice, I feel it's more important to understand the goal of the TUI mode._

**Comment by emillon** [comment]
Hi,
Thanks for your interest. I suggest opening a [discussion](https://github.com/ocaml/dune/discussions) since this is not an issue with dune and there's no action required to fix it.

**Comment by Alizter** [comment]
@emillon There is an option to convert this issue into a discussion on the side. I had an answer partially written out but I didn't finish it due to time. I'll try to finish it soon.

**Comment by favonia** [comment]
> Hi, Thanks for your interest. I suggest opening a [discussion](https://github.com/ocaml/dune/discussions) since this is not an issue with dune and there's no action required to fix it.

@emillon I think it _is_ an issue with dune: maybe the new mode should be redesigned, and at very least the problematic Unicode characters must be replaced and more documentation should be added. My concern implies that changes need to be made. However, I agree that we don't have a consensus on what exactly should be done yet, so maybe a discussion is indeed more appropriate. I believe (or hope) there will be actionable items generated from this discussion.

**Reply by emillon** [reply]
Sure, feel free to open an issue if the output looks off in your terminal. That can be fixed independently of your question.

**Reply by favonia** [reply]
Well, those Unicode characters are what I worry least about... I prefer focusing on my main question, and my point is that the question (not the output being off) _is_ an issue that dune eventually has to address. Our disagreement seems to be that you think my question can be answered without "fixing" dune, but I think we do have to change it somehow. :slightly_smiling_face:

**Reply by Alizter** [reply]
I have known about the potential difficulties East Asian character sets present for TUI, but I haven't yet got a solution to dealing with them. I will need to do some more research and ask other developers how they handle this problem. AFAIK there isn't a silver bullet but there should be many good-enough solutions. Please do create an issue specifically about it however.

Regarding the rendering of the ogham relative to a horizontal rule. I also know about this issue and I have not yet found a better solution. Whether or not this symbol aligns depends on the font the terminal emulator is using. In my testing this has been the case for around half. Other alternative characters are too clunky in my opinion, hence why I haven't changed it. I'm open to suggestions but it's not a priority right now. Please again open a specific issue for this.

**Reply by favonia** [reply]
@Alizter Hmm. I think you might have confused "East Asian character sets" with "East Asian widths." They are _very_ different. Please consult UAX 11 for detailed definitions and discussions. To clarify, my comment about East Asian widths has _nothing_ to do with "East Asian character sets."

**Reply by Alizter** [reply]
@favonia It seems I don't fully understand the issue then. I understood what you wrote to mean: "if you use box drawing characters and east asian characters then they don't align due to the ambiguous east asian width of the box drawing characters". How do the east asian character widths apply when not using those character sets?

**Reply by favonia** [reply]
East Asian widths are a property of Unicode characters that is _not_ defined only for East Asian characters. Please consult UAX 11 of the Unicode standard.

**Reply by favonia** [reply]
Forgot to say this: East Asian (legacy) character sets contain non-"East Asian" characters. Maybe that's why you were confused by the width property? Again please consult the Unicode standard---I could explain it but prefer not to spend too much time repeating the relevant parts of the standard while trying to focus on my main question (why we need the TUI mode in dune).

**Comment by Alizter** [comment]
I agree that there is not much documentation available so hopefully I can clear some things up. I've been meaning to write some documentation on TUI eventually so this might be a good place to start.

## What is the TUI mode?

The Dune TUI (terminal user interface) is intended to be an easy-to-use, portable and interactive alternative to the usual command-line usage of Dune.

## What does it do?

Today in 3.11 you can click on messages to minimise them, open up a help screen, resize your window as you like, scroll through messages, and quit using `q`.

## What are the plans for the future?

The current features seem limited, and that's OK. Hence why we marked TUI as "experimental" in the 3.11.0 change log.

The current feature I am working on that will build on TUI is the jobs tab:
- https://github.com/ocaml/dune/pull/8601

This will allow the users to view the status of processes that Dune is running and interact with them appropriately.

In order to not have such a wide feature disparity with regular Dune commands, we also wish to implement a subset of the functionality for regular the `dune monitor` command. Unfortunately this means taming a long standing issue with multiline status lines being cleared improperly, which is quite a complicated thing to do. This is what I am currently working on and the latest progress can be found in https://github.com/ocaml/dune/pull/8619.

We have all sorts of other ideas like:
- Clicking a button to promote a diff.
- Modifying build goals
- Displaying jobs (which I mentioned above)
- Navigating to errors

Each of these will have their own specific features soon, but for the time are working on them one-at-a-time. 

## How does it work?

The UI is based on the Notty library and allows us to draw images easily without worrying too much about rendering considerations within Dune. It also uses nottui from lwd which provides support for UI elements based on Notty.

In the first release of TUI back in 3.8.0, we drew the exact same thing as `--display quiet` using only Notty and with version 3.11.0 we added support for some limited interactivity to further demonstrate the utility.

## GUI frameworks have been around for years; Why do all this in the terminal?

The main answer is portability. We would like Dune to:
1. Run on just about every platform supported by OCaml.
2. Do so with no system dependencies (or an acceptable few in the case of watch mode).

It's very difficult to find a GUI framework that meets these two requirements. Any sophisticated GUI framework will have a library that needs to be depended on, and simpler frameworks tend to be lacking in cross-platform support.

There are likely two things that anybody running Dune has: A terminal and a web browser. Both of these options satisfy our constraints and we have chosen to go with the first since that's how everybody is running Dune already. It is also probably more  familiar to programmers and light weight compared to spinning up a browser.

Another advantage of having a terminal based application is that it becomes easy to use Dune via other modes of interaction such as `ssh`. This coupled with the `dune monitor` command https://github.com/ocaml/dune/pull/8152 will allow users to inspect remote Dune processes.

TUI is also very suitable for usage in a terminal multiplexer such as tmux.

## How does this interact with VSCode/OCaml Platform/LSP?

The TUI mode is not meant to replace any LSP functionality but only to compliment it. As more features get added it can pave the way for similar features to be added elsewhere.

In the future, Dune will probably become based more on RPC which means that regular users will interact with it in the command line less frequently, instead interacting via LSP to build, run tests, promote files etc. In that world, it makes sense to have a way to connect to a running Dune process which is what `dune monitor` does (and will likely be what other Dune commands do too). When this is done, it will be helpful to have a nice easy-to-use interface that is not tied to a particular LSP client.

Where TUI is really useful is for Dune/OCaml specific features that LSP does not support. This allows the feature to be implemented once for Dune rather than leaving it to the other LSP clients to additionally support.

## What are some other examples of other TUI applications?

My go to example would be lazygit https://github.com/jesseduffield/lazygit which also has some similar design constraints to us. It is essentially an interactive git client that runs in your terminal and is meant to be easy to use.

For more examples on GitHub see here: https://github.com/rothgar/awesome-tuis

**Reply by favonia** [reply]
@Alizter Thank you for your explanation. The major thing that still confused me is why we cannot have this tool created outside dune. When reading your proposal, I am imagining an independent TUI client talking to dune through the RPC protocol, not something built in `dune`. There is a difference between _creating a nice interface for a core tool_ and _directly adding that interface into the core tool_. For example, we have a nice `btop` but we did not build `btop` into the kernel or `ps`. We also have `git` but not a fancy TUI/GUI in it: `gitk` and `git gui` (and obviously `lazygit` you mentioned) are independent projects.

`dune` has become the cornerstone of many OCaml projects and its reliability and efficiency will greatly help the community thrive. So far, I have difficulties seeing why the problem you described needs a _built-in_ TUI interface. Could you possibly elaborate on that? More specifically, why can't we write an independent client `dune-tui` that achieves what you wanted? In my opinion, we should probably start by building such an independent tool, and only when there's popular need, we integrate (a good) one into the core tool. An independent tool will also verify that the RPC protocol really does not depend on a particular implementation.

PS: To clarify, I'm not against with a "simple", "primitive" TUI, e.g., like the core `git`.

PPS: _(This is more like a clarification:)_ I have personally used and even created TUI programs. The general usefulness of TUI is not what I objected to. My concerns only arise when more separable features are added to the core tools _before_ showing that they will be used by many people.

**Reply by rgrinberg** [reply]
> The major thing that still confused me is why we cannot have this tool created outside dune. 

It *can* happen outside of dune if there was demand for it and it improved the pace of development. I see that as highly unlikely. Your next point alludes to why:

> When reading your proposal, I am imagining an independent TUI client talking to dune through the RPC protocol, not something built in dune

It would be very likely that developing such a TUI client will require major additions to the RPC protocol. It's easier to do such synchronized changes in a single repository. In fact, our approach so far is to develop the UI first, and then see what RPC changes will be required to eventually to make the UI work on top of it. There's some experiments in that direction with `$ dune monitor`.

There's also a non technical reason: the authors of this TUI are all regular dune developers. They are all quite busy, and juggling another repo and syncing it with upstream dune changes would just create unnecessary busy work for them. If the team was bigger, and the TUI client was developed by non core devs, we could reconsider this point.

> For example, we have a nice btop but we did not build btop into the kernel or ps. We also have git but not a fancy TUI/GUI in it: gitk and git gui (and obviously lazygit you mentioned) are independent projects

Plenty of counter examples as well. The cmake gui exists in the same repository. I think hg comes with it's own tui for some operations.

> Could you possibly elaborate on that? More specifically, why can't we write an independent client dune-tui that achieves what you wanted? In my opinion, we should probably start by building such an independent tool, and only when there's popular need, we integrate (a good) one into the core tool. An independent tool will also verify that the RPC protocol really does not depend on a particular implementation.

As I've mentioned, I think your approach will slow development. I'd like to hear some arguments against developing TUI inside the dune repo. I see some potential downsides, but I think we've mitigated them pretty well:

1. The TUI might require additional dependencies. To address that, we've come up with a system of including dependencies in dune that does not hurt the portability of dune

2. The TUI might impact the stability of dune in general. We've been very careful to keep the UI concerns isolated from the rest of the code. Moreover, we've made the TUI opt-in rather than opt-out.

3. The TUI slows down building dune itself. This doesn't seem like a problem to me so far, but should it become a problem, we could consider splitting the UI into a separate binary. This would not require introducing a new repository.

**Reply by jonsterling** [reply]
> It can happen outside of dune if there was demand for it and it improved the pace of development

I am not feeling too strongly, but when I read this part of your comment @rgrinberg, it made me think at first that you were arguing _against_ including a TUI — indeed, why would we add one to dune if there was not demand for a TUI?

**Reply by rgrinberg** [reply]
I meant demand from potential contributors willing to work on the TUI rather than demand from users of the TUI.

**Reply by favonia** [reply]
@rgrinberg Thank you for the insight about the development cost, which I share a lot. I believe there's a way to address your concerns while satisfying mine. You have already pointed out the way: it seems the tool can be developed _in_ the same repository but as a _separate_ OPAM package depending on the same version of the private library (like `dune-glob`). This way, the hypothetical `dune-tui` and the core `dune` will be developed in lock steps, addressing the development costs. On the other hand, building the core `dune` will not include `dune-tui`, addressing my strong desire to make the core tool reliable and efficient.

To be clear, I can see the usefulness of including a server and a simple client (not a fancy TUI client) of the RPC into the _core_ dune (by that, I meant it would be linked together into the same binary). The reason is that clusters maintained in a school or a company might only have the core dune coming from the distro (e.g., an Ubuntu package), and the only way to run an RPC server on such clusters is to include the features directly in the core tool. However, I imagine that the fancy TUI client is used on a device that the user will have more control over, which means installing a separate tool should be much easier.

I understand that any change requires extra time and work. However, I'd like to see whether you and other dune developers will, in principle, object if I or someone else makes a PR to isolate the fancy TUI into a separate package. I am strongly motivated to do so, but I am also quite busy, and I prefer to avoid making a PR when the maintainers will likely reject it.

PS: @rgrinberg I have read through several GitHub issues/PRs when trying to learn what this TUI is for, and I did recognize your efforts across many issues and PRs to minimize the impact of the TUI on the rest of the codebase. I really appreciate it.

**Reply by Alizter** [reply]
>I understand that any change requires extra time and work. However, I'd like to see whether you and other dune developers will, in principle, object if I or someone else makes a PR to isolate the fancy TUI into a separate package. I am strongly motivated to do so, but I am also quite busy, and I prefer to avoid making a PR when the maintainers will likely reject it.

Could you elaborate on your motivation to separate TUI from the rest of Dune?

If I've understood what you've written (and please pardon me if I haven't) your main concern is that you believe Dune has become more inefficient or unreliable with the addition of TUI? If this is the case, could you expand on whether these are concrete observations that you have today or just worries? In the first case, that is definitely a regression that we should look into.

**Reply by favonia** [reply]
@Alizter There are already observations demonstrating the quality degradation. I agree with @rgrinberg that things still look _okay,_ and thus I'm only urging the dune team to exercise extra precaution. However, I disagree that dune is as efficient and reliable as before, and here are some concrete observations that made me worried about the current TUI mode and its future development.

# Build time

I have noticed that the CI took more and more time for a while. The CPU user time (summation of all CPU user usage, not the wall time) to build `dune` on my machine roughly grows like this when OCaml 5.1 is used along with OPAM:

   - 3.4.0: 41 seconds
   - 3.5.0: 42 seconds
   - 3.6.0: 42 seconds
   - 3.7.0: 43 seconds
   - **3.8.0: 53 seconds (~23% increase)**
   - **3.9.0: 58 seconds**
   - 3.10.0: 57 seconds
   - 3.11.0: 60 seconds

There's a huge increase in build time from 3.7 to 3.8. There is also some increase from 3.8 to 3.9, though much less significant. I didn't have time to check, and I understand that sometimes it is inevitable. However, since my attempt to learn the TUI mode, I realized that the **10 second increase in 3.8 might be strongly related to the vendoring of `notty` and `uutf`,** which together constitute more than 2300 lines (without comments or blank lines). **All of these seem to be for the "fancy" TUI.** Of course, there might be other factors slowing down the building of 3.8, but at very least it looks suspicious to me.

# Terminal and Unicode: two extremely complicated items

Terminal control sequences (especially the more advanced ones used in the fancy TUI) and Unicode are extremely complicated. So for, I have seen several issues in the current implementation:

1. **Incorrect implementation of dim/faint:** I have personally helped fix it, but the bug had been in the codebase for about ten months, and it took me quite some efforts to communicate the problem. I might not be able to do it again.
2. **Non-portable assumptions about visual widths** (mentioned in this discussion).
3. **An arguably suboptimal choice of characters for Unicode Art** (mentioned in this discussion).
4. **Usage of potentially outdated Unicode data** from Unicode 13.0.0 (this is unavoidable until notty is updated).
5. **The lack of Windows support,** also mainly because of notty.

There are countless traps and pitfalls about Unicode and terminal control sequences that are extremely subtle to avoid. Given the lack of person power mentioned by @rgrinberg, I hope the dune team could reassess whether the TUI interface is really worth the trouble. With all respect, if the dune team _really_ wants to go this way, the team might benefit from some outside help on these issues. (I'm willing to help, but I might not have enough expertise, and I am not able to track changes to the dune codebase.)

# Conclusive remark: stability for fundamental tools

Finally, I want to repeat my view on stability: I believe we should exercise extra precaution when introducing new features to a tool as fundamental as dune. I feel the idea that "we can fix it if there's a regression" could be dangerous when the entire ecosystem is depending on this tool. To clarify, I don't mean that we should not make any change, but that **precaution is a valid reason not to include _non-essential_ features in a tool that must maintain the highest standard of reliability and efficiency.** Therefore, I do not think my worries must be justified by concrete observations, though there _are_ such observations as mentioned above.

**My question still stands: why should we integrate the fancy TUI into the core dune?** (And to clarify, I think a basic TUI is essential, but the fancy TUI is not, and developing the fancy TUI in the same repository does not count as "putting it into the core dune").

**Reply by ejgallego** [reply]
I guess the core point here is that Dune still doesn't provide a good API to build externally tools like the Tui mode, or to integrate with other kind of setups. As Dune is more and more popular, this becomes a need, but also maintain APIs is very hard.

Think of the model of the Linux Kernel for example, there is lots of stuff in the tree, due to API compat, but the quality varies. It is true that the kernel has a complex configure system allowing to enable/disable several features.

On the other hand, other than the build time increase for Dune itself, I don't see how TUI would affect the quality or stability of Dune.

**Reply by ejgallego** [reply]
IMVHO , and given the lack of stable API / RPC yet, developing tools like the Coq mode or Tui in-tree are useful experiments to understand how such an API would look like.

**Reply by favonia** [reply]
@ejgallego I would not dismiss the build time increase this quickly ---I do hope the CI can be faster. As for the Linux kernel, most "fancy" things in the Linux kernel can be built as modules, a strategy I will happily agree with. In the OCaml world, that means a separate OPAM package (which can still be in the same repository). To me, what's important is not really whether they have a build configuration system to turn things on/off, but that they have an API to isolate modules (and now I understand we might still lack one). As for the quality concerns, in addition to the build time, I have explained that human resources are limited, and it isn't easy to maintain a fancy _and correct_ TUI. That said, if people want to do it anyway, I'm in no position to stop anyone. What confused me was your seeming resistance to placing the fancy TUI as a separate OPAM package. Why can't that arrangement satisfy the need to develop dune and dune-tui in lock steps?

I have been hearing the defense of the current situation as an "experiment." The defense seems to imply that once the RPC protocol and/or the API matures, all these fancy UIs will be removed from the core package (just like dune-glob). Is that the case? It's not very obvious to me from the design document presented by @Alizter.

**Reply by rgrinberg** [reply]
The build times have increased due to the 0install solver and the opam client libraries being included. Those are needed for dune's package management feature. TUI was introduced in 3.7, so by your measurements, we've increased the build time by an entire second with this feature.

The rest of your of points regarding complexity and stability are only applicable in situations where the users chooses to enable the TUI. A more convincing argument would demonstrate how stability is impacted for everyone else.

The point of implementing TUI on top of RPC isn't to gain stability (you've yet to demonstrate this point), but to allow the TUI to manage more than one running dune instance at a time, and in the future, perhaps connect to remote dune instances.

> To clarify, I don't mean that we should not make any change, but that precaution is a valid reason not to include non-essential features in a tool that must maintain the highest standard of reliability and efficiency. Therefore, I do not think my worries must be justified by concrete observations, though there are such observations as mentioned above.

The problem is that the set of "non-essential" features is different for every user. I find that even the current implementation TUI useful in my personal workflow, while other features in dune that I never use seem very much "non-essential" to me.

All in all, I remain unconvinced that introducing a separate package is a good use of development time. However, the way I like to develop dune is to cede the initiative to people who are motivated to contribute. So if you're willing to put in the work of making the UI work over RPC and in general help maintain this component of dune. I see nothing wrong with letting you decide in which package this component should live.

**Reply by ejgallego** [reply]
Why are you building dune in your CI to start with? All the CIs I have cache the whole opam env (either using Docker or opam-setup action)

Having Dune itself build fast is indeed a worthwhile goal, but also as you say, I can see it very low in terms of dev priorities. Most of use care that Dune builds projects fast, Dune build time itself is not even measured.

I agree about the desire to have an API, and about all that other concerns about how hard is to do a TUI.

I'm not sure what would change in the current setup if dune TUI was put in a different opam package, how that would make the rest of Dune better, other than having a faster build? This can be already done with a very small patch, but as far as I see it changes little.

**Reply by favonia** [reply]
@rgrinberg Thank you for the comment that a PR (if I find time) will not be wasted. However, I hope I can get some clarification on your comment about the 3.7.0 versus 3.8.0. The vendering of `notty` and `uutf` did not seem to happen in 3.7.0 because you have personally reverted the vendoring in 8a4a1136c2fcc8daf76a0226b4ec56eb03d84b09 (https://github.com/ocaml/dune/pull/6780) before the final release of 3.7.0), and the vendoring of `0install` did not seem to happen in 3.8.0, either. (Yes, that seems to have caused the 5-second increase, and I understand that might be needed, so I did not complain about it!) Did I get this part of the history wrong? The detail is very relevant here because it can be a good estimation about how much saving we could have by moving the fancy TUI into a separate package. I will be much less motivated if the difference is only one second.

**Reply by rgrinberg** [reply]
No, you're right. You should be able to get accurate information by removing TUI temporarily.

**Reply by favonia** [reply]
@ejgallego Thanks for your comments. I'm curious about the CI: we are using https://github.com/ocaml/setup-ocaml and its designer has made a deliberate choice _not_ to cache the entire opam env. I already observed that the uploading of the archived directory takes quite some time so I tend to agree with the designer (until I have time doing more experiments). Do you have evidence showing it's not the case? If so, maybe we should change how setup-ocaml works?

**Reply by favonia** [reply]
@ejgallego The long story is that we have tried the Docker approach and the cache-it-all strategy before we transitioned to setup-ocaml, and both can be annoying but I don't want to elaborate here. The existence of https://github.com/ocaml/setup-ocaml really saves us lots of maintenance efforts. It sounds that you are suggesting a more efficient script with lower maintenance, and I am very interested.

**Reply by favonia** [reply]
@rgrinberg Thinking about it, even if I manage to make a PR to make it a package, which will probably make me easier (shorter build time), I probably would not have time to _maintain_ it, at least not as actively as @Alizter. I'm still holding the view that it's better to make the tool a separate package, but I'm also following the same principle of "people who contribute most, decide". I have said enough in this discussion and perhaps I should leave it as it is now. @emillon @ejgallego Thank you all for the conversation.
